File: javaserver/src/com/streamserver/util/SMessageStream.java
package com.streamserver.util;
import java.util.ArrayList;
import java.lang.RuntimeException;
import java.util.List;
import java.util.LinkedList;
import java.util.Iterator;
// this class is the container of message
// message can be an byte array of int , short or other format
// so message is big endian, when you create message data, it must be big endian
public class SMessageStream
{
    public static class SMessage 
    {
        public byte[] data;
    }
    private static class SMessagePacket
    {
        public byte[] data;
        public int len;
        public int offset;	
    }
    public static final int S_NO_ERROR = 0;
    public static final int S_WAIT_MORE = 1;
    public SMessageStream()
    {
    }
    public int getNextMessage(SMessage outMessage)
    {
        int ret = S_WAIT_MORE;
	byte[] outData = null;
	if(outMessage == null)
	    throw new RuntimeException("out message is null");
	synchronized(mMessagePacketList)
	{
	    if(!mMessagePacketList.isEmpty())
	    {
	        int dataLen = getDataLen();
		System.out.println("### data len = " + dataLen);
	        if(dataLen != 0)
		{
		    outData = new byte[dataLen];
		    int[] offsetBackup = new int[mMessagePacketList.size()];
                    backupOffset(offsetBackup);
		    int leftDataLen = dataLen;
		    int outDataIndex = 0;
		    int i = 0;
		    Iterator<SMessagePacket> it = mMessagePacketList.iterator();
		    //for(int i = 0 ; i < mMessagePacketList.size() ; i+=)
		    while(it.hasNext())
		    {
		        SMessagePacket mp = it.next();//mMessagePacketList.get(i);
			int packetLen = mp.len - mp.offset;
			if(packetLen >= leftDataLen)
			{
			    System.arraycopy(mp.data, mp.offset, outData, outDataIndex, leftDataLen);
			    offsetBackup[i] += leftDataLen;
			    leftDataLen = 0;
			    break;
			}
			else
			{
			    System.arraycopy(mp.data, mp.offset, outData, outDataIndex, packetLen);
			    leftDataLen -= packetLen;
			    offsetBackup[i] += packetLen;
			    outDataIndex += packetLen;
			}
			i++;
		    }
		    if(leftDataLen == 0)
		    {
		        ret = S_NO_ERROR;
		        changeOffset(offsetBackup);	
		    }
		    else
		    {
		        ret = S_WAIT_MORE;
		    }
                    //delete data packet from message list
		    List<SMessagePacket> newMessagePacketList = new LinkedList<SMessagePacket>();
		    it = mMessagePacketList.iterator();
		    while(it.hasNext())
		    {
		        SMessagePacket mp = it.next();
			if(mp.offset < mp.len)
			{
		            newMessagePacketList.add(mp);
			}
		    }
		    mMessagePacketList.clear();
		    mMessagePacketList = newMessagePacketList;
		    //end
		}
	        	
	    }
	}
	if(ret == S_NO_ERROR)
	{
	    outMessage.data = outData; 
	}
	return ret;
    }
    private void backupOffset(int[] offsets)
    {
        Iterator<SMessagePacket> it = mMessagePacketList.iterator();
        int i = 0;
        //for(int i = 0 ; i < mMessagePacketList.size() ; i++)
	while(it.hasNext())
	{
	    SMessagePacket mp = it.next();
	    offsets[i] = mp.offset;//mMessagePacketList.get(i).offset;
	    i++;
	}
    }
    private void changeOffset(int[] offsets)
    {
	Iterator<SMessagePacket> it = mMessagePacketList.iterator();
	int i = 0;
        //for(int i = 0 ; i < mMessagePacketList.size() ; i++)
	while(it.hasNext())
	{
	    SMessagePacket mp = it.next();
	    mp.offset = offsets[i];
	    i++;
	}

    }
    public int addMessagePacket(byte[] data)
    {
        if(data == null)
	    throw new RuntimeException("data is null pointer");
	SMessagePacket mp = new SMessagePacket();
	mp.data = data;
	mp.len = data.length;
	mp.offset = 0;
	synchronized(mMessagePacketList)
	{
	    mMessagePacketList.add(mp);
	}
	return S_NO_ERROR;
    }
    public int getMessagePacketCount()
    {
        int count = 0 ; 
	synchronized(mMessagePacketList)
	{
            count = mMessagePacketList.size();
	}
	return count;
    }
    private int getDataLen()
    {
	Iterator<SMessagePacket> it = mMessagePacketList.iterator();
        SMessagePacket front = null;
	if(it.hasNext())
	    front = it.next();
	if(front.offset >= front.len)
	    throw new RuntimeException("offset is greater than len");
	int leftDataLen = front.len - front.offset;
	short dataLen = 0;
	if(leftDataLen >= 3)
	{
	    byte[] tmpData = new byte[2];
	    System.arraycopy(front.data, front.offset + 1, tmpData, 0, 2);
	    int high = tmpData[0] & 0xff;
	    int low = tmpData[1] & 0xff;
	    dataLen = (short)((high << 8) | low);
	    return dataLen;
	}
	else
	{
	    byte[] headerData = new byte[3];
	    int currentLen = front.len - front.offset;
	    System.arraycopy(front, front.offset, headerData, 0, currentLen);
            leftDataLen = 3 - currentLen;
	    it = mMessagePacketList.iterator();
	    it.next();
	    //for(int i = 1 ; i < mMessagePacketList.size() ; i++)
	    while(it.hasNext())
	    {
	        SMessagePacket mp = it.next();
		if(mp.offset >= mp.len)
		    throw new RuntimeException("offset is greater than len");
		int packetCurrLen = mp.len - mp.offset;
		if(leftDataLen <= packetCurrLen)
		{
		    System.arraycopy(mp.data, mp.offset, headerData, currentLen, leftDataLen);
		    break;
		}
		else
		{
		    System.arraycopy(mp.data, mp.offset, headerData, currentLen, packetCurrLen);
		    leftDataLen -= packetCurrLen;
		    currentLen += packetCurrLen;
		}
	    }
	    if(leftDataLen == 0)
	    {
                byte[] tmpData = new byte[2];
                System.arraycopy(headerData, 1, tmpData, 0, 2);
                int high = tmpData[0] & 0xff;
  	        int low = tmpData[1] & 0xff;
	        dataLen = (short)((high << 8) | low);	
	        return dataLen;	
	    }
	    else
	    {
	        return 0;
	    }
	}
    }

    private List<SMessagePacket> mMessagePacketList = new LinkedList<SMessagePacket>(); 
}

File: core/SMessageStream.cpp
#include "SMessageStream.h"
#include "SMutex.h"
#include "SUtil.h"
#include "SLog.h"
#include <list>
#include <vector>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
//////////////////////////////////

//////////////////////////////////
class SMessageStream::SMessageStreamImpl
{
public:
    ~SMessageStreamImpl();
    typedef list<SMessagePacket*> SMessagePacketList;
    int getDataLen();
    void backupOffset(vector<int>& offsetBackup);
    void changeOffset(vector<int>& offsetBackup);
    SMessagePacketList mMessagePacketList;
    SMutex mMessagePacketListMutex;
};
SMessageStream::SMessageStreamImpl::~SMessageStreamImpl()
{
    SAutoMutex mutex(&mMessagePacketListMutex);
    SMessagePacketList::iterator it;
    for(it = mMessagePacketList.begin() ; it != mMessagePacketList.end() ; it++)
    {
        delete *it;
    } 
}
int SMessageStream::SMessageStreamImpl::getDataLen()
{
    SMessagePacket* front = mMessagePacketList.front();
    unsigned char* dataStart = front->mData + front->offset;
    SASSERT(front->offset < front->len);
    uint16_t dataLen = 0;
    int leftDataLen = front->len - front->offset;
    if(leftDataLen >= 3)
    {
        memcpy(&dataLen, dataStart + 1, 2);
        dataLen = SUtil::Net2HostInt16(dataLen);
        return dataLen;
    }
    else
    {
        unsigned char headerData[3];
        unsigned char* p = headerData;
        int currentLen = front->len - front->offset;
        memcpy(headerData, dataStart, currentLen);
        leftDataLen = 3 - (front->len - front->offset);
        p += currentLen;
        SMessagePacketList::iterator it = mMessagePacketList.begin();
        it++;
        for(; it != mMessagePacketList.end() ; it++)
        {
            SMessagePacket* mp = *it;
            SASSERT(mp->offset < mp->len);
            currentLen = mp->len - mp->offset;
            if(leftDataLen <= currentLen)
            {
                memcpy(p, mp->mData + mp->offset, leftDataLen);
                leftDataLen = 0;
                break;
            }
            else
            {
                memcpy(p, mp->mData + mp->offset, currentLen);
                leftDataLen -= currentLen;
                p += currentLen;
            }
        }
        if(leftDataLen == 0)
        {
            memcpy(&dataLen, headerData + 1, 2);
            dataLen = SUtil::Net2HostInt16(dataLen);
        }
        return dataLen;
    }

}
void SMessageStream::SMessageStreamImpl::backupOffset(vector<int>& offsetBackup)
{
    SMessagePacketList::iterator it;
    int i;
    for(it = mMessagePacketList.begin(), i = 0; it != mMessagePacketList.end(); it++, i++)
    {
        offsetBackup[i] = (*it)->offset;
    }
}
void SMessageStream::SMessageStreamImpl::changeOffset(vector<int>& offsetBackup)
{
    SMessagePacketList::iterator it;
    int i;
    for(it = mMessagePacketList.begin(), i = 0; it != mMessagePacketList.end(); it++, i++)
    {
        (*it)->offset = offsetBackup[i];
    }

}
//////////////////////////////////////
static bool canDelete(const SMessagePacket* mp)
{
    return mp->isAllConsumed();
}
///////////////////////////////////////
SMessageStream::SMessageStream() : mImpl(new SMessageStreamImpl)
{
    
}
SMessageStream::~SMessageStream()
{}
int SMessageStream::getNextMessage(SMessage* out)
{
    /*
    if(mHead == NULL)
        return WAIT_MORE;
    SMessagePacket* p = mHead;
    unsigned char* dataStart = p->mData + offset;
    if(dataStart == (p->mData + len))    
    {
        mHead = p->mNext;
    }
    */
    int ret = S_WAIT_MORE;
    SAutoMutex mutex(&mImpl->mMessagePacketListMutex);
    if(mImpl->mMessagePacketList.empty())
        return ret;
    try
    {
        uint16_t dataLen = mImpl->getDataLen();
        if(dataLen == 0)
        {
            ret = S_WAIT_MORE;
            return ret;
        }
        uint8_t* outData = new uint8_t[dataLen];
        uint8_t* outDataHead = outData;
        vector<int> offsetBackup(mImpl->mMessagePacketList.size(), 0);
        mImpl->backupOffset(offsetBackup);
        SMessageStreamImpl::SMessagePacketList::iterator it = mImpl->mMessagePacketList.begin();
        int leftDataLen = dataLen;
        unsigned char* dataStart = NULL;
        int i = 0;
        for(; it != mImpl->mMessagePacketList.end(); it++, i++)
        {
            SMessagePacket* mp = *it;
            int currentLen = mp->len - mp->offset;
            dataStart = mp->mData + mp->offset;
            if(currentLen >= leftDataLen)
            {
                memcpy(outData, dataStart, leftDataLen);
                //mp->offset += leftDataLen;
                offsetBackup[i] += leftDataLen;
                leftDataLen = 0;
                break;
            }
            else
            {
                memcpy(outData, dataStart, currentLen);
                leftDataLen -= currentLen;
                //mp->offset += currentLen;
                offsetBackup[i] += currentLen;
                outData += currentLen;
            }
        }
        if(leftDataLen == 0)
        {
            ret = S_NO_ERROR;
            out->data = outDataHead;
            out->len = dataLen;
            mImpl->changeOffset(offsetBackup);
        }
        else
        {
            ret = S_WAIT_MORE;
        }
        //delete message packet
        SMessageStreamImpl::SMessagePacketList deleteMessagePackets;
        for(it = mImpl->mMessagePacketList.begin(); it != mImpl->mMessagePacketList.end(); it++)
        {
            SMessagePacket* data = *it;
            if(data->isAllConsumed())
            {
                deleteMessagePackets.push_back(data);
            }
        }
        mImpl->mMessagePacketList.remove_if(canDelete);
        for(it = deleteMessagePackets.begin() ; it != deleteMessagePackets.end(); it++)
        {
            SMessagePacket* data = *it;
            delete data;
        }
        //end
    }
    catch(...)
    {

    }
    return ret;
}
int SMessageStream::addMessagePacket(unsigned char* data, int len, bool own)
{
    SASSERT(data != NULL && len > 0);
    SMessagePacket* newPacket = new SMessagePacket;
    newPacket->mOwn = own;
    if(!own)
    {
        newPacket->mData = new unsigned char[len];
        memcpy(newPacket->mData, data, len);
    }
    else
    {
        newPacket->mData = data;
    }
    newPacket->len = len;
    //mImpl->mMessagePacketListMutex.lock();
    SAutoMutex mutex(&mImpl->mMessagePacketListMutex);
    try
    {
        mImpl->mMessagePacketList.push_back(newPacket);
    }
    catch(...)
    {

    }
    //mImpl->mMessagePacketListMutex.unlock();
    return S_NO_ERROR;
}
int SMessageStream::getMessagePacketCount()
{
    int count = 0;
    //mImpl->mMessagePacketListMutex.lock();
    SAutoMutex mutex(&mImpl->mMessagePacketListMutex);
    count = mImpl->mMessagePacketList.size();
    //mImpl->mMessagePacketListMutex.unlock();
    return count;
}
void SMessageStream::mapMessagePacket(SMessagePacketFunctor& functor, bool clearPacketList)
{
    //SLog::msg("map lock start\n");
    SAutoMutex mutex(&mImpl->mMessagePacketListMutex);
    //SLog::msg("map lock end\n");
    SMessageStreamImpl::SMessagePacketList::iterator it;
    for(it = mImpl->mMessagePacketList.begin() ; it != mImpl->mMessagePacketList.end(); it++)
    {
        SMessagePacket* smp = *it;
        //SLog::msg("### output packe %p ####\n", smp);
        functor.handleMessagePacket(smp);
    }
    if(clearPacketList)
    {
        for(it = mImpl->mMessagePacketList.begin() ; it != mImpl->mMessagePacketList.end(); it++)
        {
            SMessagePacket* smp = *it;
            delete smp;
        }
        mImpl->mMessagePacketList.clear();
    }
}
